// @sinclair/typebox/build/esm/type/type/javascript.mjs
import { JsonTypeBuilder } from '@sinclair/typebox/build/esm/type/type/json.mjs';
import { Argument } from '@sinclair/typebox/build/esm/type/argument/index.mjs';
import { AsyncIterator } from '@sinclair/typebox/build/esm/type/async-iterator/index.mjs';
import { Awaited } from '@sinclair/typebox/build/esm/type/awaited/index.mjs';
import { BigInt } from '@sinclair/typebox/build/esm/type/bigint/index.mjs';
import { Constructor } from '@sinclair/typebox/build/esm/type/constructor/index.mjs';
import { ConstructorParameters } from '@sinclair/typebox/build/esm/type/constructor-parameters/index.mjs';
import { Date } from '@sinclair/typebox/build/esm/type/date/index.mjs';
import { Function as FunctionType } from '@sinclair/typebox/build/esm/type/function/index.mjs';
import { InstanceType } from '@sinclair/typebox/build/esm/type/instance-type/index.mjs';
import { Instantiate } from '@sinclair/typebox/build/esm/type/instantiate/index.mjs';
import { Iterator } from '@sinclair/typebox/build/esm/type/iterator/index.mjs';
import { Parameters } from '@sinclair/typebox/build/esm/type/parameters/index.mjs';
import { Promise } from '@sinclair/typebox/build/esm/type/promise/index.mjs';
import { RegExp } from '@sinclair/typebox/build/esm/type/regexp/index.mjs';
import { ReturnType } from '@sinclair/typebox/build/esm/type/return-type/index.mjs';
import { Symbol } from '@sinclair/typebox/build/esm/type/symbol/index.mjs';
import { Uint8Array } from '@sinclair/typebox/build/esm/type/uint8array/index.mjs';
import { Undefined } from '@sinclair/typebox/build/esm/type/undefined/index.mjs';
import { Void } from '@sinclair/typebox/build/esm/type/void/index.mjs';
/** JavaScript Type Builder with Static Resolution for TypeScript */
export class JavaScriptTypeBuilder extends JsonTypeBuilder {
    /** `[JavaScript]` Creates a Generic Argument Type */
    Argument(index) {
        return Argument(index);
    }
    /** `[JavaScript]` Creates a AsyncIterator type */
    AsyncIterator(items, options) {
        return AsyncIterator(items, options);
    }
    /** `[JavaScript]` Constructs a type by recursively unwrapping Promise types */
    Awaited(schema, options) {
        return Awaited(schema, options);
    }
    /** `[JavaScript]` Creates a BigInt type */
    BigInt(options) {
        return BigInt(options);
    }
    /** `[JavaScript]` Extracts the ConstructorParameters from the given Constructor type */
    ConstructorParameters(schema, options) {
        return ConstructorParameters(schema, options);
    }
    /** `[JavaScript]` Creates a Constructor type */
    Constructor(parameters, instanceType, options) {
        return Constructor(parameters, instanceType, options);
    }
    /** `[JavaScript]` Creates a Date type */
    Date(options = {}) {
        return Date(options);
    }
    /** `[JavaScript]` Creates a Function type */
    Function(parameters, returnType, options) {
        return FunctionType(parameters, returnType, options);
    }
    /** `[JavaScript]` Extracts the InstanceType from the given Constructor type */
    InstanceType(schema, options) {
        return InstanceType(schema, options);
    }
    /** `[JavaScript]` Instantiates a type with the given parameters */
    Instantiate(schema, parameters) {
        return Instantiate(schema, parameters);
    }
    /** `[JavaScript]` Creates an Iterator type */
    Iterator(items, options) {
        return Iterator(items, options);
    }
    /** `[JavaScript]` Extracts the Parameters from the given Function type */
    Parameters(schema, options) {
        return Parameters(schema, options);
    }
    /** `[JavaScript]` Creates a Promise type */
    Promise(item, options) {
        return Promise(item, options);
    }
    /** `[JavaScript]` Creates a RegExp type */
    RegExp(unresolved, options) {
        return RegExp(unresolved, options);
    }
    /** `[JavaScript]` Extracts the ReturnType from the given Function type */
    ReturnType(type, options) {
        return ReturnType(type, options);
    }
    /** `[JavaScript]` Creates a Symbol type */
    Symbol(options) {
        return Symbol(options);
    }
    /** `[JavaScript]` Creates a Undefined type */
    Undefined(options) {
        return Undefined(options);
    }
    /** `[JavaScript]` Creates a Uint8Array type */
    Uint8Array(options) {
        return Uint8Array(options);
    }
    /** `[JavaScript]` Creates a Void type */
    Void(options) {
        return Void(options);
    }
}
