// @sinclair/typebox/build/esm/type/omit/omit.mjs
import { CreateType } from '@sinclair/typebox/build/esm/type/create/type.mjs';
import { Discard } from '@sinclair/typebox/build/esm/type/discard/discard.mjs';
import { TransformKind } from '@sinclair/typebox/build/esm/type/symbols/symbols.mjs';
import { Computed } from '@sinclair/typebox/build/esm/type/computed/index.mjs';
import { Literal } from '@sinclair/typebox/build/esm/type/literal/index.mjs';
import { IndexPropertyKeys } from '@sinclair/typebox/build/esm/type/indexed/index.mjs';
import { Intersect } from '@sinclair/typebox/build/esm/type/intersect/index.mjs';
import { Union } from '@sinclair/typebox/build/esm/type/union/index.mjs';
import { Object } from '@sinclair/typebox/build/esm/type/object/index.mjs';
// ------------------------------------------------------------------
// Mapped
// ------------------------------------------------------------------
import { OmitFromMappedKey } from '@sinclair/typebox/build/esm/type/omit/omit-from-mapped-key.mjs';
import { OmitFromMappedResult } from '@sinclair/typebox/build/esm/type/omit/omit-from-mapped-result.mjs';
// ------------------------------------------------------------------
// TypeGuard
// ------------------------------------------------------------------
import { IsMappedKey, IsIntersect, IsUnion, IsObject, IsSchema, IsMappedResult, IsLiteralValue, IsRef } from '@sinclair/typebox/build/esm/type/guard/kind.mjs';
import { IsArray as IsArrayValue } from '@sinclair/typebox/build/esm/type/guard/value.mjs';
// prettier-ignore
function FromIntersect(types, propertyKeys) {
    return types.map((type) => OmitResolve(type, propertyKeys));
}
// prettier-ignore
function FromUnion(types, propertyKeys) {
    return types.map((type) => OmitResolve(type, propertyKeys));
}
// ------------------------------------------------------------------
// FromProperty
// ------------------------------------------------------------------
// prettier-ignore
function FromProperty(properties, key) {
    const { [key]: _, ...R } = properties;
    return R;
}
// prettier-ignore
function FromProperties(properties, propertyKeys) {
    return propertyKeys.reduce((T, K2) => FromProperty(T, K2), properties);
}
// prettier-ignore
function FromObject(properties, propertyKeys) {
    const options = Discard(properties, [TransformKind, '$id', 'required', 'properties']);
    const omittedProperties = FromProperties(properties['properties'], propertyKeys);
    return Object(omittedProperties, options);
}
// prettier-ignore
function UnionFromPropertyKeys(propertyKeys) {
    const result = propertyKeys.reduce((result, key) => IsLiteralValue(key) ? [...result, Literal(key)] : result, []);
    return Union(result);
}
// prettier-ignore
function OmitResolve(properties, propertyKeys) {
    return (IsIntersect(properties) ? Intersect(FromIntersect(properties.allOf, propertyKeys)) :
        IsUnion(properties) ? Union(FromUnion(properties.anyOf, propertyKeys)) :
            IsObject(properties) ? FromObject(properties, propertyKeys) :
                Object({}));
}
/** `[Json]` Constructs a type whose keys are picked from the given type */
// prettier-ignore
export function Omit(type, key, options) {
    const typeKey = IsArrayValue(key) ? UnionFromPropertyKeys(key) : key;
    const propertyKeys = IsSchema(key) ? IndexPropertyKeys(key) : key;
    const isTypeRef = IsRef(type);
    const isKeyRef = IsRef(key);
    return (IsMappedResult(type) ? OmitFromMappedResult(type, propertyKeys, options) :
        IsMappedKey(key) ? OmitFromMappedKey(type, key, options) :
            (isTypeRef && isKeyRef) ? Computed('Omit', [type, typeKey], options) :
                (!isTypeRef && isKeyRef) ? Computed('Omit', [type, typeKey], options) :
                    (isTypeRef && !isKeyRef) ? Computed('Omit', [type, typeKey], options) :
                        CreateType({ ...OmitResolve(type, propertyKeys), ...options }));
}
