// pure-rand/lib/esm/pure-rand-default.js
import { congruential32 } from 'pure-rand/lib/esm/generator/LinearCongruential.js';
import { mersenne } from 'pure-rand/lib/esm/generator/MersenneTwister.js';
import { xorshift128plus } from 'pure-rand/lib/esm/generator/XorShift.js';
import { xoroshiro128plus } from 'pure-rand/lib/esm/generator/XoroShiro.js';
import { uniformArrayIntDistribution } from 'pure-rand/lib/esm/distribution/UniformArrayIntDistribution.js';
import { uniformBigIntDistribution } from 'pure-rand/lib/esm/distribution/UniformBigIntDistribution.js';
import { uniformIntDistribution } from 'pure-rand/lib/esm/distribution/UniformIntDistribution.js';
import { unsafeUniformArrayIntDistribution } from 'pure-rand/lib/esm/distribution/UnsafeUniformArrayIntDistribution.js';
import { unsafeUniformBigIntDistribution } from 'pure-rand/lib/esm/distribution/UnsafeUniformBigIntDistribution.js';
import { unsafeUniformIntDistribution } from 'pure-rand/lib/esm/distribution/UnsafeUniformIntDistribution.js';
import { skipN } from 'pure-rand/lib/esm/distribution/SkipN.js';
import { generateN } from 'pure-rand/lib/esm/distribution/GenerateN.js';
import { unsafeGenerateN } from 'pure-rand/lib/esm/distribution/UnsafeGenerateN.js';
import { unsafeSkipN } from 'pure-rand/lib/esm/distribution/UnsafeSkipN.js';
var __type = 'module';
var __version = '7.0.1';
var __commitHash = '2248506b66d969d1a8b477a4dde8e24cbac33e6a';
export { __type, __version, __commitHash, generateN, skipN, unsafeGenerateN, unsafeSkipN, congruential32, mersenne, xorshift128plus, xoroshiro128plus, uniformArrayIntDistribution, uniformBigIntDistribution, uniformIntDistribution, unsafeUniformArrayIntDistribution, unsafeUniformBigIntDistribution, unsafeUniformIntDistribution, };
