// @sinclair/typebox/build/esm/type/awaited/awaited.mjs
import { CreateType } from '@sinclair/typebox/build/esm/type/create/type.mjs';
import { Computed } from '@sinclair/typebox/build/esm/type/computed/index.mjs';
import { Intersect } from '@sinclair/typebox/build/esm/type/intersect/index.mjs';
import { Union } from '@sinclair/typebox/build/esm/type/union/index.mjs';
import { Ref } from '@sinclair/typebox/build/esm/type/ref/index.mjs';
// ------------------------------------------------------------------
// TypeGuard
// ------------------------------------------------------------------
import { IsIntersect, IsUnion, IsPromise, IsRef, IsComputed } from '@sinclair/typebox/build/esm/type/guard/kind.mjs';
// prettier-ignore
function FromComputed(target, parameters) {
    return Computed('Awaited', [Computed(target, parameters)]);
}
// prettier-ignore
function FromRef($ref) {
    return Computed('Awaited', [Ref($ref)]);
}
// prettier-ignore
function FromIntersect(types) {
    return Intersect(FromRest(types));
}
// prettier-ignore
function FromUnion(types) {
    return Union(FromRest(types));
}
// prettier-ignore
function FromPromise(type) {
    return Awaited(type);
}
// prettier-ignore
function FromRest(types) {
    return types.map(type => Awaited(type));
}
/** `[JavaScript]` Constructs a type by recursively unwrapping Promise types */
export function Awaited(type, options) {
    return CreateType(IsComputed(type) ? FromComputed(type.target, type.parameters) : IsIntersect(type) ? FromIntersect(type.allOf) : IsUnion(type) ? FromUnion(type.anyOf) : IsPromise(type) ? FromPromise(type.item) : IsRef(type) ? FromRef(type.$ref) : type, options);
}
