// @sinclair/typebox/build/cjs/type/type/json.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonTypeBuilder = void 0;
const index_1 = require("@sinclair/typebox/build/cjs/type/any");
const index_2 = require("@sinclair/typebox/build/cjs/type/array");
const index_3 = require("@sinclair/typebox/build/cjs/type/boolean");
const index_4 = require("@sinclair/typebox/build/cjs/type/composite");
const index_5 = require("@sinclair/typebox/build/cjs/type/const");
const index_6 = require("@sinclair/typebox/build/cjs/type/enum");
const index_7 = require("@sinclair/typebox/build/cjs/type/exclude");
const index_8 = require("@sinclair/typebox/build/cjs/type/extends");
const index_9 = require("@sinclair/typebox/build/cjs/type/extract");
const index_10 = require("@sinclair/typebox/build/cjs/type/indexed");
const index_11 = require("@sinclair/typebox/build/cjs/type/integer");
const index_12 = require("@sinclair/typebox/build/cjs/type/intersect");
const index_13 = require("@sinclair/typebox/build/cjs/type/intrinsic");
const index_14 = require("@sinclair/typebox/build/cjs/type/keyof");
const index_15 = require("@sinclair/typebox/build/cjs/type/literal");
const index_16 = require("@sinclair/typebox/build/cjs/type/mapped");
const index_17 = require("@sinclair/typebox/build/cjs/type/never");
const index_18 = require("@sinclair/typebox/build/cjs/type/not");
const index_19 = require("@sinclair/typebox/build/cjs/type/null");
const index_20 = require("@sinclair/typebox/build/cjs/type/module");
const index_21 = require("@sinclair/typebox/build/cjs/type/number");
const index_22 = require("@sinclair/typebox/build/cjs/type/object");
const index_23 = require("@sinclair/typebox/build/cjs/type/omit");
const index_24 = require("@sinclair/typebox/build/cjs/type/optional");
const index_25 = require("@sinclair/typebox/build/cjs/type/partial");
const index_26 = require("@sinclair/typebox/build/cjs/type/pick");
const index_27 = require("@sinclair/typebox/build/cjs/type/readonly");
const index_28 = require("@sinclair/typebox/build/cjs/type/readonly-optional");
const index_29 = require("@sinclair/typebox/build/cjs/type/record");
const index_30 = require("@sinclair/typebox/build/cjs/type/recursive");
const index_31 = require("@sinclair/typebox/build/cjs/type/ref");
const index_32 = require("@sinclair/typebox/build/cjs/type/required");
const index_33 = require("@sinclair/typebox/build/cjs/type/rest");
const index_34 = require("@sinclair/typebox/build/cjs/type/string");
const index_35 = require("@sinclair/typebox/build/cjs/type/template-literal");
const index_36 = require("@sinclair/typebox/build/cjs/type/transform");
const index_37 = require("@sinclair/typebox/build/cjs/type/tuple");
const index_38 = require("@sinclair/typebox/build/cjs/type/union");
const index_39 = require("@sinclair/typebox/build/cjs/type/unknown");
const index_40 = require("@sinclair/typebox/build/cjs/type/unsafe");
/** Json Type Builder with Static Resolution for TypeScript */
class JsonTypeBuilder {
    // ------------------------------------------------------------------------
    // Modifiers
    // ------------------------------------------------------------------------
    /** `[Json]` Creates a Readonly and Optional property */
    ReadonlyOptional(type) {
        return (0, index_28.ReadonlyOptional)(type);
    }
    /** `[Json]` Creates a Readonly property */
    Readonly(type, enable) {
        return (0, index_27.Readonly)(type, enable ?? true);
    }
    /** `[Json]` Creates a Optional property */
    Optional(type, enable) {
        return (0, index_24.Optional)(type, enable ?? true);
    }
    // ------------------------------------------------------------------------
    // Types
    // ------------------------------------------------------------------------
    /** `[Json]` Creates an Any type */
    Any(options) {
        return (0, index_1.Any)(options);
    }
    /** `[Json]` Creates an Array type */
    Array(items, options) {
        return (0, index_2.Array)(items, options);
    }
    /** `[Json]` Creates a Boolean type */
    Boolean(options) {
        return (0, index_3.Boolean)(options);
    }
    /** `[Json]` Intrinsic function to Capitalize LiteralString types */
    Capitalize(schema, options) {
        return (0, index_13.Capitalize)(schema, options);
    }
    /** `[Json]` Creates a Composite object type */
    Composite(schemas, options) {
        return (0, index_4.Composite)(schemas, options); // (error) TS 5.4.0-dev - review TComposite implementation
    }
    /** `[JavaScript]` Creates a readonly const type from the given value. */
    Const(value, options) {
        return (0, index_5.Const)(value, options);
    }
    /** `[Json]` Creates a Enum type */
    Enum(item, options) {
        return (0, index_6.Enum)(item, options);
    }
    /** `[Json]` Constructs a type by excluding from unionType all union members that are assignable to excludedMembers */
    Exclude(unionType, excludedMembers, options) {
        return (0, index_7.Exclude)(unionType, excludedMembers, options);
    }
    /** `[Json]` Creates a Conditional type */
    Extends(L, R, T, F, options) {
        return (0, index_8.Extends)(L, R, T, F, options);
    }
    /** `[Json]` Constructs a type by extracting from type all union members that are assignable to union */
    Extract(type, union, options) {
        return (0, index_9.Extract)(type, union, options);
    }
    /** `[Json]` Returns an Indexed property type for the given keys */
    Index(type, key, options) {
        return (0, index_10.Index)(type, key, options);
    }
    /** `[Json]` Creates an Integer type */
    Integer(options) {
        return (0, index_11.Integer)(options);
    }
    /** `[Json]` Creates an Intersect type */
    Intersect(types, options) {
        return (0, index_12.Intersect)(types, options);
    }
    /** `[Json]` Creates a KeyOf type */
    KeyOf(type, options) {
        return (0, index_14.KeyOf)(type, options);
    }
    /** `[Json]` Creates a Literal type */
    Literal(literalValue, options) {
        return (0, index_15.Literal)(literalValue, options);
    }
    /** `[Json]` Intrinsic function to Lowercase LiteralString types */
    Lowercase(type, options) {
        return (0, index_13.Lowercase)(type, options);
    }
    /** `[Json]` Creates a Mapped object type */
    Mapped(key, map, options) {
        return (0, index_16.Mapped)(key, map, options);
    }
    /** `[Json]` Creates a Type Definition Module. */
    Module(properties) {
        return (0, index_20.Module)(properties);
    }
    /** `[Json]` Creates a Never type */
    Never(options) {
        return (0, index_17.Never)(options);
    }
    /** `[Json]` Creates a Not type */
    Not(type, options) {
        return (0, index_18.Not)(type, options);
    }
    /** `[Json]` Creates a Null type */
    Null(options) {
        return (0, index_19.Null)(options);
    }
    /** `[Json]` Creates a Number type */
    Number(options) {
        return (0, index_21.Number)(options);
    }
    /** `[Json]` Creates an Object type */
    Object(properties, options) {
        return (0, index_22.Object)(properties, options);
    }
    /** `[Json]` Constructs a type whose keys are omitted from the given type */
    Omit(schema, selector, options) {
        return (0, index_23.Omit)(schema, selector, options);
    }
    /** `[Json]` Constructs a type where all properties are optional */
    Partial(type, options) {
        return (0, index_25.Partial)(type, options);
    }
    /** `[Json]` Constructs a type whose keys are picked from the given type */
    Pick(type, key, options) {
        return (0, index_26.Pick)(type, key, options);
    }
    /** `[Json]` Creates a Record type */
    Record(key, value, options) {
        return (0, index_29.Record)(key, value, options);
    }
    /** `[Json]` Creates a Recursive type */
    Recursive(callback, options) {
        return (0, index_30.Recursive)(callback, options);
    }
    /** `[Json]` Creates a Ref type. The referenced type must contain a $id */
    Ref(...args) {
        return (0, index_31.Ref)(args[0], args[1]);
    }
    /** `[Json]` Constructs a type where all properties are required */
    Required(type, options) {
        return (0, index_32.Required)(type, options);
    }
    /** `[Json]` Extracts interior Rest elements from Tuple, Intersect and Union types */
    Rest(type) {
        return (0, index_33.Rest)(type);
    }
    /** `[Json]` Creates a String type */
    String(options) {
        return (0, index_34.String)(options);
    }
    /** `[Json]` Creates a TemplateLiteral type */
    TemplateLiteral(unresolved, options) {
        return (0, index_35.TemplateLiteral)(unresolved, options);
    }
    /** `[Json]` Creates a Transform type */
    Transform(type) {
        return (0, index_36.Transform)(type);
    }
    /** `[Json]` Creates a Tuple type */
    Tuple(types, options) {
        return (0, index_37.Tuple)(types, options);
    }
    /** `[Json]` Intrinsic function to Uncapitalize LiteralString types */
    Uncapitalize(type, options) {
        return (0, index_13.Uncapitalize)(type, options);
    }
    /** `[Json]` Creates a Union type */
    Union(types, options) {
        return (0, index_38.Union)(types, options);
    }
    /** `[Json]` Creates an Unknown type */
    Unknown(options) {
        return (0, index_39.Unknown)(options);
    }
    /** `[Json]` Creates a Unsafe type that will infers as the generic argument T */
    Unsafe(options) {
        return (0, index_40.Unsafe)(options);
    }
    /** `[Json]` Intrinsic function to Uppercase LiteralString types */
    Uppercase(schema, options) {
        return (0, index_13.Uppercase)(schema, options);
    }
}
exports.JsonTypeBuilder = JsonTypeBuilder;
