// @sinclair/typebox/build/cjs/type/type/type.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.Rest = exports.Required = exports.RegExp = exports.Ref = exports.Recursive = exports.Record = exports.ReadonlyOptional = exports.Readonly = exports.Promise = exports.Pick = exports.Partial = exports.Parameters = exports.Optional = exports.Omit = exports.Object = exports.Number = exports.Null = exports.Not = exports.Never = exports.Module = exports.Mapped = exports.Literal = exports.KeyOf = exports.Iterator = exports.Uppercase = exports.Lowercase = exports.Uncapitalize = exports.Capitalize = exports.Intersect = exports.Integer = exports.Instantiate = exports.InstanceType = exports.Index = exports.Function = exports.Extract = exports.Extends = exports.Exclude = exports.Enum = exports.Date = exports.ConstructorParameters = exports.Constructor = exports.Const = exports.Composite = exports.Boolean = exports.BigInt = exports.Awaited = exports.AsyncIterator = exports.Array = exports.Argument = exports.Any = void 0;
exports.Void = exports.Unsafe = exports.Unknown = exports.Union = exports.Undefined = exports.Uint8Array = exports.Tuple = exports.Transform = exports.TemplateLiteral = exports.Symbol = exports.String = exports.ReturnType = void 0;
// ------------------------------------------------------------------
// Type: Module
// ------------------------------------------------------------------
var index_1 = require("@sinclair/typebox/build/cjs/type/any");
Object.defineProperty(exports, "Any", { enumerable: true, get: function () { return index_1.Any; } });
var index_2 = require("@sinclair/typebox/build/cjs/type/argument");
Object.defineProperty(exports, "Argument", { enumerable: true, get: function () { return index_2.Argument; } });
var index_3 = require("@sinclair/typebox/build/cjs/type/array");
Object.defineProperty(exports, "Array", { enumerable: true, get: function () { return index_3.Array; } });
var index_4 = require("@sinclair/typebox/build/cjs/type/async-iterator");
Object.defineProperty(exports, "AsyncIterator", { enumerable: true, get: function () { return index_4.AsyncIterator; } });
var index_5 = require("@sinclair/typebox/build/cjs/type/awaited");
Object.defineProperty(exports, "Awaited", { enumerable: true, get: function () { return index_5.Awaited; } });
var index_6 = require("@sinclair/typebox/build/cjs/type/bigint");
Object.defineProperty(exports, "BigInt", { enumerable: true, get: function () { return index_6.BigInt; } });
var index_7 = require("@sinclair/typebox/build/cjs/type/boolean");
Object.defineProperty(exports, "Boolean", { enumerable: true, get: function () { return index_7.Boolean; } });
var index_8 = require("@sinclair/typebox/build/cjs/type/composite");
Object.defineProperty(exports, "Composite", { enumerable: true, get: function () { return index_8.Composite; } });
var index_9 = require("@sinclair/typebox/build/cjs/type/const");
Object.defineProperty(exports, "Const", { enumerable: true, get: function () { return index_9.Const; } });
var index_10 = require("@sinclair/typebox/build/cjs/type/constructor");
Object.defineProperty(exports, "Constructor", { enumerable: true, get: function () { return index_10.Constructor; } });
var index_11 = require("@sinclair/typebox/build/cjs/type/constructor-parameters");
Object.defineProperty(exports, "ConstructorParameters", { enumerable: true, get: function () { return index_11.ConstructorParameters; } });
var index_12 = require("@sinclair/typebox/build/cjs/type/date");
Object.defineProperty(exports, "Date", { enumerable: true, get: function () { return index_12.Date; } });
var index_13 = require("@sinclair/typebox/build/cjs/type/enum");
Object.defineProperty(exports, "Enum", { enumerable: true, get: function () { return index_13.Enum; } });
var index_14 = require("@sinclair/typebox/build/cjs/type/exclude");
Object.defineProperty(exports, "Exclude", { enumerable: true, get: function () { return index_14.Exclude; } });
var index_15 = require("@sinclair/typebox/build/cjs/type/extends");
Object.defineProperty(exports, "Extends", { enumerable: true, get: function () { return index_15.Extends; } });
var index_16 = require("@sinclair/typebox/build/cjs/type/extract");
Object.defineProperty(exports, "Extract", { enumerable: true, get: function () { return index_16.Extract; } });
var index_17 = require("@sinclair/typebox/build/cjs/type/function");
Object.defineProperty(exports, "Function", { enumerable: true, get: function () { return index_17.Function; } });
var index_18 = require("@sinclair/typebox/build/cjs/type/indexed");
Object.defineProperty(exports, "Index", { enumerable: true, get: function () { return index_18.Index; } });
var index_19 = require("@sinclair/typebox/build/cjs/type/instance-type");
Object.defineProperty(exports, "InstanceType", { enumerable: true, get: function () { return index_19.InstanceType; } });
var index_20 = require("@sinclair/typebox/build/cjs/type/instantiate");
Object.defineProperty(exports, "Instantiate", { enumerable: true, get: function () { return index_20.Instantiate; } });
var index_21 = require("@sinclair/typebox/build/cjs/type/integer");
Object.defineProperty(exports, "Integer", { enumerable: true, get: function () { return index_21.Integer; } });
var index_22 = require("@sinclair/typebox/build/cjs/type/intersect");
Object.defineProperty(exports, "Intersect", { enumerable: true, get: function () { return index_22.Intersect; } });
var index_23 = require("@sinclair/typebox/build/cjs/type/intrinsic");
Object.defineProperty(exports, "Capitalize", { enumerable: true, get: function () { return index_23.Capitalize; } });
Object.defineProperty(exports, "Uncapitalize", { enumerable: true, get: function () { return index_23.Uncapitalize; } });
Object.defineProperty(exports, "Lowercase", { enumerable: true, get: function () { return index_23.Lowercase; } });
Object.defineProperty(exports, "Uppercase", { enumerable: true, get: function () { return index_23.Uppercase; } });
var index_24 = require("@sinclair/typebox/build/cjs/type/iterator");
Object.defineProperty(exports, "Iterator", { enumerable: true, get: function () { return index_24.Iterator; } });
var index_25 = require("@sinclair/typebox/build/cjs/type/keyof");
Object.defineProperty(exports, "KeyOf", { enumerable: true, get: function () { return index_25.KeyOf; } });
var index_26 = require("@sinclair/typebox/build/cjs/type/literal");
Object.defineProperty(exports, "Literal", { enumerable: true, get: function () { return index_26.Literal; } });
var index_27 = require("@sinclair/typebox/build/cjs/type/mapped");
Object.defineProperty(exports, "Mapped", { enumerable: true, get: function () { return index_27.Mapped; } });
var index_28 = require("@sinclair/typebox/build/cjs/type/module");
Object.defineProperty(exports, "Module", { enumerable: true, get: function () { return index_28.Module; } });
var index_29 = require("@sinclair/typebox/build/cjs/type/never");
Object.defineProperty(exports, "Never", { enumerable: true, get: function () { return index_29.Never; } });
var index_30 = require("@sinclair/typebox/build/cjs/type/not");
Object.defineProperty(exports, "Not", { enumerable: true, get: function () { return index_30.Not; } });
var index_31 = require("@sinclair/typebox/build/cjs/type/null");
Object.defineProperty(exports, "Null", { enumerable: true, get: function () { return index_31.Null; } });
var index_32 = require("@sinclair/typebox/build/cjs/type/number");
Object.defineProperty(exports, "Number", { enumerable: true, get: function () { return index_32.Number; } });
var index_33 = require("@sinclair/typebox/build/cjs/type/object");
Object.defineProperty(exports, "Object", { enumerable: true, get: function () { return index_33.Object; } });
var index_34 = require("@sinclair/typebox/build/cjs/type/omit");
Object.defineProperty(exports, "Omit", { enumerable: true, get: function () { return index_34.Omit; } });
var index_35 = require("@sinclair/typebox/build/cjs/type/optional");
Object.defineProperty(exports, "Optional", { enumerable: true, get: function () { return index_35.Optional; } });
var index_36 = require("@sinclair/typebox/build/cjs/type/parameters");
Object.defineProperty(exports, "Parameters", { enumerable: true, get: function () { return index_36.Parameters; } });
var index_37 = require("@sinclair/typebox/build/cjs/type/partial");
Object.defineProperty(exports, "Partial", { enumerable: true, get: function () { return index_37.Partial; } });
var index_38 = require("@sinclair/typebox/build/cjs/type/pick");
Object.defineProperty(exports, "Pick", { enumerable: true, get: function () { return index_38.Pick; } });
var index_39 = require("@sinclair/typebox/build/cjs/type/promise");
Object.defineProperty(exports, "Promise", { enumerable: true, get: function () { return index_39.Promise; } });
var index_40 = require("@sinclair/typebox/build/cjs/type/readonly");
Object.defineProperty(exports, "Readonly", { enumerable: true, get: function () { return index_40.Readonly; } });
var index_41 = require("@sinclair/typebox/build/cjs/type/readonly-optional");
Object.defineProperty(exports, "ReadonlyOptional", { enumerable: true, get: function () { return index_41.ReadonlyOptional; } });
var index_42 = require("@sinclair/typebox/build/cjs/type/record");
Object.defineProperty(exports, "Record", { enumerable: true, get: function () { return index_42.Record; } });
var index_43 = require("@sinclair/typebox/build/cjs/type/recursive");
Object.defineProperty(exports, "Recursive", { enumerable: true, get: function () { return index_43.Recursive; } });
var index_44 = require("@sinclair/typebox/build/cjs/type/ref");
Object.defineProperty(exports, "Ref", { enumerable: true, get: function () { return index_44.Ref; } });
var index_45 = require("@sinclair/typebox/build/cjs/type/regexp");
Object.defineProperty(exports, "RegExp", { enumerable: true, get: function () { return index_45.RegExp; } });
var index_46 = require("@sinclair/typebox/build/cjs/type/required");
Object.defineProperty(exports, "Required", { enumerable: true, get: function () { return index_46.Required; } });
var index_47 = require("@sinclair/typebox/build/cjs/type/rest");
Object.defineProperty(exports, "Rest", { enumerable: true, get: function () { return index_47.Rest; } });
var index_48 = require("@sinclair/typebox/build/cjs/type/return-type");
Object.defineProperty(exports, "ReturnType", { enumerable: true, get: function () { return index_48.ReturnType; } });
var index_49 = require("@sinclair/typebox/build/cjs/type/string");
Object.defineProperty(exports, "String", { enumerable: true, get: function () { return index_49.String; } });
var index_50 = require("@sinclair/typebox/build/cjs/type/symbol");
Object.defineProperty(exports, "Symbol", { enumerable: true, get: function () { return index_50.Symbol; } });
var index_51 = require("@sinclair/typebox/build/cjs/type/template-literal");
Object.defineProperty(exports, "TemplateLiteral", { enumerable: true, get: function () { return index_51.TemplateLiteral; } });
var index_52 = require("@sinclair/typebox/build/cjs/type/transform");
Object.defineProperty(exports, "Transform", { enumerable: true, get: function () { return index_52.Transform; } });
var index_53 = require("@sinclair/typebox/build/cjs/type/tuple");
Object.defineProperty(exports, "Tuple", { enumerable: true, get: function () { return index_53.Tuple; } });
var index_54 = require("@sinclair/typebox/build/cjs/type/uint8array");
Object.defineProperty(exports, "Uint8Array", { enumerable: true, get: function () { return index_54.Uint8Array; } });
var index_55 = require("@sinclair/typebox/build/cjs/type/undefined");
Object.defineProperty(exports, "Undefined", { enumerable: true, get: function () { return index_55.Undefined; } });
var index_56 = require("@sinclair/typebox/build/cjs/type/union");
Object.defineProperty(exports, "Union", { enumerable: true, get: function () { return index_56.Union; } });
var index_57 = require("@sinclair/typebox/build/cjs/type/unknown");
Object.defineProperty(exports, "Unknown", { enumerable: true, get: function () { return index_57.Unknown; } });
var index_58 = require("@sinclair/typebox/build/cjs/type/unsafe");
Object.defineProperty(exports, "Unsafe", { enumerable: true, get: function () { return index_58.Unsafe; } });
var index_59 = require("@sinclair/typebox/build/cjs/type/void");
Object.defineProperty(exports, "Void", { enumerable: true, get: function () { return index_59.Void; } });
